### PR TITLE
Feature/variadic intersection type

### LINKED ISSUE #183: Union for variadic types
There are a lot of situations may occur when we need to define union from several cases.
For example: need to define enum of states.

This would look like:
```php
use Psl\Type as T;

$codec = T\shape([
    // ... rest of fields
    'state' => T\union(
        T\literal_scalar('NEW'),
        T\union(
            T\literal_scalar('APPROVED'),
            T\union(
                T\literal_scalar('REJECTED'),
                T\union(
                    T\literal_scalar('COMPLETED'),
                    T\union(
                        T\literal_scalar('FROZEN'),
                        T\literal_scalar('ERROR')
                    )
                )
            )
        )
    ),
]);
```

Inferred type would be:
```
array{id: string, state: "APPROVED"|"COMPLETED"|"ERROR"|"FROZEN"|"NEW"|"REJECTED"}
```

---

I guess it would be great to have better solution like:
```php
use Psl\Type as T;

$codec = T\shape([
    'id' => T\string(),
    'state' => T\union_of(
        T\literal_scalar('NEW'),
        T\literal_scalar('APPROVED'),
        T\literal_scalar('REJECTED'),
        T\literal_scalar('COMPLETED'),
        T\literal_scalar('FROZEN'),
        T\literal_scalar('ERROR'),
    ),
]);
```

I do understand that `Psl\Type\Internal\UnionType` has only 2 branches: **left** and **right**
That's why propose just implement a function with signature:
```php
/**
 * @template T
 *
 * @param TypeInterface<T> $left_type
 * @param TypeInterface<T> $right_type
 * @param TypeInterface<T> ...$others
 * @return TypeInterface<T>
 *
 * @no-named-arguments
 */
function union_of(
    TypeInterface $left_type,
    TypeInterface $right_type,
    TypeInterface ...$others
): TypeInterface { // actually Psl\Type\Internal\UnionType returns here
    ...
}
```

---

As alternatives I may guess there is more suitable name for function.
This may be named like `literals` or `one_of` or etc.

This function may also be implemented in user land, but I think it would be great to have unified solution inside core of the library.

---

Want to hear other guys thoughts.
If this would be approved, I may produce PR.
As I can see, this is very light enhancement.

### PR DESCRIPTION
**Pull request by nzour**
Union variadic types issue #183 and PR #184

### CODE REVIEW

File: tests/static-analysis/Type/intersection.php [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+            Type\object(stdClass::class),
+        ),
+    );
+
+    /** @psalm-suppress MissingThrowsDocblock */
+    $new_codec = Type\intersection(
+        Type\object(Map::class),
+        Type\object(ResultInterface::class),
+        Type\object(stdClass::class),
+    );
```
Comment (azjezz) [inline tests/static-analysis/Type/intersection.php]: hm, can you test this with 4 objects? i don't think it's going to work....

File: tests/static-analysis/Type/intersection.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
+<?php
+
+declare(strict_types=1);
+
+use Psl\Collection\Map;
+use Psl\Result\ResultInterface;
+use Psl\Type;
+
+/**
+ * @psalm-suppress UnusedParam
```
Comment (azjezz) [inline tests/static-analysis/Type/intersection.php:11]: p.s: you can rename the argument to `$_value` and psalm won't annoy you about unused argument ;)

File: tests/static-analysis/Type/intersection.php (hunk 1 above)
Comment (azjezz) [reply tests/static-analysis/Type/intersection.php]: hm, it seems to work here https://psalm.dev/r/433e9e351b, not sure why i thought it wouldn't ...

**Comment by azjezz** [comment]
could you please add a unit test to create an intersection with more than 2 types? ( to keep code coverage the same )

File: tests/static-analysis/Type/intersection.php (hunk 1 above)
Comment (nzour) [reply tests/static-analysis/Type/intersection.php]: Sure, also check https://psalm.dev/r/e1d0238b41

**Comment by nzour** [comment]
> could you please add a unit test to create an intersection with more than 2 types? ( to keep code coverage the same )

Yep, I'll do it

File: tests/unit/Type/IntersectionTypeTest.php [hunk 3]
```diff
@@ -53,14 +53,36 @@ public function getToStringExamples(): iterable
...
+            '(Psl\Collection\CollectionInterface|Iterator)&Psl\Collection\IndexAccessInterface'
+        ];
+
+        yield [
+            Type\intersection(
+                Type\object(IndexAccessInterface::class),
+                Type\object(CollectionInterface::class),
+                Type\object(Iterator::class),
+                Type\shape(['id' => Type\string()]),
+            ),
```
Comment (azjezz) [inline tests/unit/Type/IntersectionTypeTest.php:84]: technically an impossible type, but okay :p

**Review by azjezz — APPROVED** [review]

**Comment by azjezz** [comment]
Thank you @zour228 !

**Pull request merged**